### ISSUE TITLE
changes in CRIU

### ISSUE THREAD

**Issue by ashwani29** [checkpoint-restore/criu#1182]
1) I don't want to run the container after restoring it, as soon as it restores the state, the container should immediately be paused, what and where should i make the changes in source code?

2) i've to keep track what PID of the processes in the checkpointed state got which PID on restoration on the destination? have to save this information. How do i do it?

**Comment by adrianreber** [comment]
>     1. I don't want to run the container after restoring it, as soon as it restores the state, the container should immediately be paused, what and where should i make the changes in source code?

CRIU supports that:  `--leave-stopped`. It depends which container runtime you are using. The container runtime must just use the existing CRIU functionality.

>     2. i've to keep track what PID of the processes in the checkpointed state got which PID on restoration on the destination? have to save this information. How do i do it?

PID will always be the same during restore. CRIU does never changed the PID of a restored process.

**Comment by ashwani29** [comment]
> > ```
> >1. I don't want to run the container after restoring it, as soon as it restores the state, the container should immediately be paused, what and where should i make the changes in source code?
> CRIU supports that: `--leave-stopped`. It depends which container runtime you are using. The container runtime must just use the existing CRIU functionality.
> 
Sorry, I've asked the question in wrong way. what i meant to say was when i migrated the checkpointed state of a container to a destination, and when restoring it using `runc restore` command, what changes should i make such that as soon as container state is restored, container on the destination side itself stops/paused? It should not run even for a ms.
> > ```
> > 2. i've to keep track what PID of the processes in the checkpointed state got which PID on restoration on the destination? have to save this information. How do i do it?
> > ```
> 
> PID will always be the same during restore. CRIU does never changed the PID of a restored process.

But when i do `runc list` the pid shown on the destination is different than what it was there on source. and also what if that particular pid is not available on destination? when the container is checkpointed on source the pstree.img files captures the information of all processes inside a container like it's pid and ppid etc. and when this checkpointed state is restored on destination side, these processes will be created on destination and will get new pids, So i've to save the information of which pid got what pid on destination?

**Comment by adrianreber** [comment]
> > > ```
> > > 1. I don't want to run the container after restoring it, as soon as it restores the state, the container should immediately be paused, what and where should i make the changes in source code?
> > > ```
> > 
> > 
> > CRIU supports that: `--leave-stopped`. It depends which container runtime you are using. The container runtime must just use the existing CRIU functionality.
> 
> Sorry, I've asked the question in wrong way. what i meant to say was when i migrated the checkpointed state of a container to a destination, and when restoring it using `runc restore` command, what changes should i make such that as soon as container state is restored, container on the destination side itself stops/paused? It should not run even for a ms.

I think I understood your question. CRIU supports that with `--leave-stopped`. You have to extend `runc` so that it tells CRIU to leave the processes stopped after restoring. This requires a code change in `runc`.

> 
> > > ```
> > > 2. i've to keep track what PID of the processes in the checkpointed state got which PID on restoration on the destination? have to save this information. How do i do it?
> > > ```
> > 
> > 
> > PID will always be the same during restore. CRIU does never changed the PID of a restored process.
> 
> But when i do `runc list` the pid shown on the destination is different than what it was there on source. and also what if that particular pid is not available on destination? when the container is checkpointed on source the pstree.img files captures the information of all processes inside a container like it's pid and ppid etc. and when this checkpointed state is restored on destination side, these processes will be created on destination and will get new pids, So i've to save the information of which pid got what pid on destination?

The PID stays the same in the PID namespace of the container. Outside of the PID namespace the PIDs will be whatever is available next. For the process only the PID in the PID namespace is relevant.

**Comment by ashwani29** [comment]
Thanks, now i'll look into runc code first and try to do it. If some problem occurs will ask here only.
and for the answer to my second part of the question, how can I look into that PID namespace when the container is stopped/paused? actually, what my goal is whenever a new process is created on the source, I've to add that process to the same parent in the container on destination side too?

**Comment by ashwani29** [comment]
If only memory image like pages and register information of the new process is added to the container on the destination, does it will imply that the new process is created in the container?

**Comment by adrianreber** [comment]
I do not understand your question. Can you rephrase?

**Comment by ashwani29** [comment]
Ok so, a container is running on the source and a replica of the container is there on destination(in paused state). As a new process joins/created in the container on the source side, my aim is to create the same process in the container on the destination side too to the same parent as it was on the source. So, if a process clones/ forks a child on the source, how can I create/ clone a child to the same parent on destination container too? This is the main goal.

So I was asking if I update the destination container with the same memory state(memory pages, register state) as of the source one will I be able to see a new process in the destination container?

Other solution I was thinking to use nsenter to enter that PID namespace and then execute a program in there which will create a child to the same parent but then that program itself will take a pid in the container namespace, created processes might not get same pid as in the source container and I will lost the namespace equivalence of two containers.

**Comment by adrianreber** [comment]
If you could describe the goal of your project it would probably be easier to understand what you want to achieve. It sounds like you want to create something like a hot-standby-container which mirrors a running container. Is it something like that you want to achieve?

> Other solution I was thinking to use nsenter to enter that PID namespace and then execute a program in there which will create a child to the same parent but then that program itself will take a pid in the container namespace, created processes might not get same pid as in the source container and I will lost the namespace equivalence of two containers.

Not sure I understand that, but you can do nsenter and then run CRIU to restore a process with the same PID in that PID namespace. CRIU can also be told to restore a process into an existing PID namespace. If you want to create a new process in an existing PID namespace you can also use the same interfaces CRIU uses to create a process with a certain PID (ns_last_pid or clone3()).

It would really help if you could give us the bigger picture of what you are trying to achieve here.

**Comment by ashwani29** [comment]
Yes , this is what i' want to achieve, my aim is to maintain a hot-standby-container which mirrors a running container on other host and the replica is passive in nature. 
> If you could describe the goal of your project it would probably be easier to understand what you want to achieve. It sounds like you want to create something like a hot-standby-container which mirrors a running container. Is it something like that you want to achieve?
> It would really help if you could give us the bigger picture of what you are trying to achieve here.

**Comment by adrianreber** [comment]
> Yes , this is what i' want to achieve, my aim is to maintain a hot-standby-container which mirrors a running container on other host and the replica is passive in nature.

Okay. Sounds like an interesting idea.

Your current approach seems to focus on using the existing CRIU code and I think this will not work without changes directly to CRIU. `pre-dump` seems to be pretty close to what you need, but `pre-dump` does not do a full dump. You would need a version of CRIU which uses the kernel's soft-dirty-tracking to do a full checkpoint without dumping all the memory pages. This way the transferred checkpoint could be still relatively small and the transfer to the destination could be reasonably fast.

From my point of view this requires code changes to combine `pre-dump` with `dump` to make use of soft-dirty-tracking in the normal `dump` mode. This is currently not possible but sounds doable. This could then all be integrated into runc to leave the restored contained in a stopped state and only start it if the main container is unreachable or goes down. I would say definitely doable but it requires code changes from your side to implement it.

**Comment by ashwani29** [comment]
I was thinking of applying changes of the checkpoint image between two consecutive checkpoints to the restored container on backup. This way it'll be highly synchronous to the container on source side and the backup container will be ready to go as soon as source one fails.

I tried to use nsenter to enter the restored container namespace to trace the PID of all the processes in the container and call clone to create a child from one of the parent process in the container but don't know how to use nsenter to run a custom program, everywhere on the internet they just used simple example like ps and ls to run inside a namespace but it is not helping. Do u have any idea how can i acheive this?

**Comment by adrianreber** [comment]
From my point of view this will not work without CRIU code changes. CRIU already has the ability to restore a process into an existing network and PID namespace.

> I tried to use nsenter to enter the restored container namespace to trace the PID of all the processes in the container and call clone to create a child from one of the parent process in the container but don't know how to use nsenter to run a custom program, everywhere on the internet they just used simple example like ps and ls to run inside a namespace but it is not helping. Do u have any idea how can i acheive this?

I do not think this will work. Just re-creating the process tree does not sound like something that will work. This requires development effort in CRIU.

It is also not clear how you will make sure that the backup container is updated with the latest version of the process on the source side. Once a process is restored CRIU, in its current form, is not really able to modify an existing process.

From my point of view CRIU is able to do the things you want it do to and the whole idea does not sound doable. I think this needs a different approach how to do things and it definitely needs code changes. Maybe https://github.com/twosigma/fastfreeze is something which would make more sense for you.

**Comment by ashwani29** [comment]
agreed, recreating process tree doesn't achieve the goal , I'll have to do all the things as criu restore does, step by step like resolving shared memory between the new processes and restored one and also among the new processes, forking the tree and rest others for each differential checkpoint.

In this link https://criu.org/Tree_after_restore, under section **Doing execv() in CRIU**, what can be done with `custom execv`. what I know for `execv` system call, it'll replace the process image with a new one. will this execv runs after the whole tree is restored or in the starting? Is there any more information related to it? any example of using it in `criu`.

**Comment by Snorch** [comment]
For better understanding of fastfreeze you can take a look on this video from LPC:

Fast checkpointing with criu-image-streamer: https://youtu.be/fSyr_IXM21Y?t=7762

**Comment by ashwani29** [comment]
@adrianreber  https://criu.org/Memory_changes_tracking this link discusses that incremental dump is possible with `--track-mem` and `prev/images/dir` even with `dump` command.

> Your current approach seems to focus on using the existing CRIU code and I think this will not work without changes directly to CRIU. `pre-dump` seems to be pretty close to what you need, but `pre-dump` does not do a full dump. You would need a version of CRIU which uses the kernel's soft-dirty-tracking to do a full checkpoint without dumping all the memory pages. This way the transferred checkpoint could be still relatively small and the transfer to the destination could be reasonably fast.
> 
> From my point of view this requires code changes to combine `pre-dump` with `dump` to make use of soft-dirty-tracking in the normal `dump` mode. This is currently not possible but sounds doable. This could then all be integrated into runc to leave the restored contained in a stopped state and only start it if the main container is unreachable or goes down. I would say definitely doable but it requires code changes from your side to implement it.
